File: .oldC++/kernel/src/ai/agents_manager.cpp
#include "jarvis/ai/agents_manager.h"
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>
#include <QJsonDocument>
#include <QJsonArray>
#include <QDateTime>
#include <algorithm>

namespace jarvis::ai {

static std::vector<std::string> jsonToStringVec(const std::string& json) {
    std::vector<std::string> result;
    auto arr = QJsonDocument::fromJson(QByteArray::fromStdString(json)).array();
    for (const auto& v : arr) {
        result.push_back(v.toString().toStdString());
    }
    return result;
}

static std::string vecToJson(const std::vector<std::string>& vec) {
    QJsonArray arr;
    for (const auto& s : vec) {
        arr.append(QString::fromStdString(s));
    }
    return QJsonDocument(arr).toJson(QJsonDocument::Compact).toStdString();
}

static Agent rowToAgent(QSqlQuery& q) {
    Agent a;
    a.id = q.value("id").toString().toStdString();
    a.name = q.value("name").toString().toStdString();
    a.description = q.value("description").toString().toStdString();
    a.model = q.value("model").toString().toStdString();
    a.systemPrompt = q.value("system_prompt").toString().toStdString();
    a.temperature = q.value("temperature").toDouble();
    a.maxTokens = q.value("max_tokens").toInt();
    a.specialty = q.value("specialty").toString().toStdString();
    a.tools = jsonToStringVec(q.value("tools").toString().toStdString());
    a.isDefault = q.value("is_default").toBool();
    a.canOrchestrate = q.value("can_orchestrate").toBool();
    a.priority = q.value("priority").toInt();
    a.createdAt = q.value("created_at").toString().toStdString();
    a.updatedAt = q.value("updated_at").toString().toStdString();
    return a;
}

class AgentsManager : public IAgentsManager {
public:
    explicit AgentsManager(const std::string& dbPath)
        : dbPath_(dbPath)
    {
        initDb();
    }

    std::vector<Agent> listAgents() override {
        std::vector<Agent> result;
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.exec("SELECT * FROM agents ORDER BY priority DESC, name ASC");
        while (q.next()) {
            result.push_back(rowToAgent(q));
        }
        return result;
    }

    Agent getAgent(const std::string& id) override {
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.prepare("SELECT * FROM agents WHERE id = ?");
        q.addBindValue(QString::fromStdString(id));
        if (q.exec() && q.next()) {
            return rowToAgent(q);
        }
        return {};
    }

    Agent createAgent(const CreateAgentDTO& dto) override {
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.prepare(R"(
            INSERT INTO agents (name, description, model, system_prompt, temperature,
                                max_tokens, specialty, tools, can_orchestrate, priority)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        )");
        q.addBindValue(QString::fromStdString(dto.name));
        q.addBindValue(QString::fromStdString(dto.description));
        q.addBindValue(QString::fromStdString(dto.model));
        q.addBindValue(QString::fromStdString(dto.systemPrompt));
        q.addBindValue(dto.temperature);
        q.addBindValue(dto.maxTokens);
        q.addBindValue(QString::fromStdString(dto.specialty));
        q.addBindValue(QString::fromStdString(vecToJson(dto.tools)));
        q.addBindValue(dto.canOrchestrate ? 1 : 0);
        q.addBindValue(dto.priority);
        if (!q.exec()) {
            throw std::runtime_error("Failed to create agent: " +
                                     q.lastError().text().toStdString());
        }

        // Fetch the newly created agent
        auto id = q.lastInsertId().toString().toStdString();
        return getAgent(id);
    }

    Agent updateAgent(const std::string& id, const CreateAgentDTO& dto) override {
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.prepare(R"(
            UPDATE agents SET
                name = ?, description = ?, model = ?, system_prompt = ?,
                temperature = ?, max_tokens = ?, specialty = ?, tools = ?,
                can_orchestrate = ?, priority = ?,
                updated_at = ?
            WHERE id = ?
        )");
        q.addBindValue(QString::fromStdString(dto.name));
        q.addBindValue(QString::fromStdString(dto.description));
        q.addBindValue(QString::fromStdString(dto.model));
        q.addBindValue(QString::fromStdString(dto.systemPrompt));
        q.addBindValue(dto.temperature);
        q.addBindValue(dto.maxTokens);
        q.addBindValue(QString::fromStdString(dto.specialty));
        q.addBindValue(QString::fromStdString(vecToJson(dto.tools)));
        q.addBindValue(dto.canOrchestrate ? 1 : 0);
        q.addBindValue(dto.priority);
        q.addBindValue(QDateTime::currentDateTimeUtc().toString(Qt::ISODate));
        q.addBindValue(QString::fromStdString(id));
        if (!q.exec()) {
            throw std::runtime_error("Failed to update agent: " +
                                     q.lastError().text().toStdString());
        }
        return getAgent(id);
    }

    bool deleteAgent(const std::string& id) override {
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.prepare("DELETE FROM agents WHERE id = ?");
        q.addBindValue(QString::fromStdString(id));
        return q.exec();
    }

    bool setDefaultAgent(const std::string& id) override {
        QSqlDatabase db = QSqlDatabase::database("ai-agents");
        db.transaction();

        QSqlQuery q(db);
        // Clear current default
        q.exec("UPDATE agents SET is_default = 0, updated_at = '" +
               QDateTime::currentDateTimeUtc().toString(Qt::ISODate) + "'");

        // Set new default
        q.prepare("UPDATE agents SET is_default = 1, updated_at = ? WHERE id = ?");
        q.addBindValue(QDateTime::currentDateTimeUtc().toString(Qt::ISODate));
        q.addBindValue(QString::fromStdString(id));
        bool ok = q.exec();

        if (ok) db.commit();
        else db.rollback();

        return ok;
    }

    Agent getDefaultAgent() override {
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.exec("SELECT * FROM agents WHERE is_default = 1 LIMIT 1");
        if (q.next()) {
            return rowToAgent(q);
        }
        // Fallback to first agent
        q.exec("SELECT * FROM agents LIMIT 1");
        if (q.next()) {
            return rowToAgent(q);
        }
        return {};
    }

    std::vector<Agent> getOrchestrationPool() override {
        std::vector<Agent> result;
        QSqlQuery q(QSqlDatabase::database("ai-agents"));
        q.exec("SELECT * FROM agents WHERE can_orchestrate = 1 ORDER BY priority DESC, name ASC");
        while (q.next()) {
            result.push_back(rowToAgent(q));
        }
        return result;
    }

private:
    std::string dbPath_;

    void initDb() {
        QString connName = "ai-agents";
        QSqlDatabase db;
        if (QSqlDatabase::contains(connName)) {
            db = QSqlDatabase::database(connName);
        } else {
            db = QSqlDatabase::addDatabase("QSQLITE", connName);
        }
        db.setDatabaseName(QString::fromStdString(dbPath_));
        if (!db.isOpen()) db.open();
        QSqlQuery q(db);
        q.exec(R"(
            CREATE TABLE IF NOT EXISTS agents (
                id TEXT PRIMARY KEY DEFAULT (lower(hex(randomblob(16)))),
                name TEXT NOT NULL,
                description TEXT NOT NULL DEFAULT '',
                model TEXT NOT NULL,
                system_prompt TEXT NOT NULL DEFAULT '',
                temperature REAL NOT NULL DEFAULT 0.7,
                max_tokens INTEGER NOT NULL DEFAULT 2048,
                specialty TEXT NOT NULL DEFAULT 'general',
                tools TEXT NOT NULL DEFAULT '[]',
                is_default INTEGER NOT NULL DEFAULT 0,
                can_orchestrate INTEGER NOT NULL DEFAULT 1,
                priority INTEGER NOT NULL DEFAULT 5,
                created_at TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%fZ', 'now')),
                updated_at TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))
            )
        )");
        // Seed default agents if table is empty
        q.exec("SELECT COUNT(*) AS cnt FROM agents");
        if (q.next() && q.value("cnt").toInt() == 0) {
            q.exec(R"(
                INSERT INTO agents (name, description, model, system_prompt, temperature, max_tokens, specialty, is_default, can_orchestrate, priority)
                VALUES ('Assistant Geral', 'Assistente geral para tarefas do dia a dia', 'llama3.2:3b',
                        'Voce e o JARVIS, um assistente de IA util, amigavel e preciso. Responda em portugues brasileiro. Use markdown para formatar respostas.',
                        0.7, 2048, 'general', 1, 1, 5)
            )");
            q.exec(R"(
                INSERT INTO agents (name, description, model, system_prompt, temperature, max_tokens, specialty, can_orchestrate, priority)
                VALUES ('Code Expert', 'Especialista em revisao de codigo e arquitetura', 'codellama:7b',
                        'Voce e um engenheiro de software senior especializado em C++, arquitetura limpa e design patterns. Revise codigo, sugira melhorias e explique conceitos complexos.',
                        0.2, 4096, 'code', 1, 8)
            )");
        }
    }
};

IAgentsManager* createAgentsManager(const std::string& dbPath) {
    return new AgentsManager(dbPath);
}

} // namespace jarvis::ai

File: backend/jarvis/agents_manager.py
import datetime
import json
import secrets
import uuid
from dataclasses import dataclass, field

from jarvis.database import Database


@dataclass
class CreateAgentDTO:
    name: str = ""
    description: str = ""
    model: str = ""
    system_prompt: str = ""
    temperature: float = 0.7
    max_tokens: int = 2048
    specialty: str = "general"
    tools: list[str] = field(default_factory=list)
    can_orchestrate: bool = True
    priority: int = 5
    is_builtin: bool = False


@dataclass
class Agent:
    id: str = ""
    name: str = ""
    description: str = ""
    model: str = ""
    system_prompt: str = ""
    temperature: float = 0.0
    max_tokens: int = 0
    specialty: str = ""
    tools: list[str] = field(default_factory=list)
    is_default: bool = False
    can_orchestrate: bool = False
    priority: int = 0
    is_builtin: bool = False
    created_at: str = ""
    updated_at: str = ""


def _row_to_agent(row) -> Agent:
    try:
        is_builtin = bool(row["is_builtin"])
    except (KeyError, IndexError):
        is_builtin = False
    return Agent(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        model=row["model"],
        system_prompt=row["system_prompt"],
        temperature=row["temperature"],
        max_tokens=row["max_tokens"],
        specialty=row["specialty"],
        tools=json.loads(row["tools"]) if row["tools"] else [],
        is_default=bool(row["is_default"]),
        can_orchestrate=bool(row["can_orchestrate"]),
        priority=row["priority"],
        is_builtin=is_builtin,
        created_at=row["created_at"],
        updated_at=row["updated_at"],
    )


# ---------------------------------------------------------------------------
# Built-in software development agents (inserted on every startup via upsert)
# ---------------------------------------------------------------------------
_BUILTIN_AGENTS = [
    {
        "id": "a1000000-0000-0000-0000-000000000001",
        "name": "Orquestrador Principal",
        "description": "Coordenador geral do fluxo de desenvolvimento de software. Analisa a entrada do usuário, distribui tarefas entre os agentes especializados e consolida os resultados.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Orquestrador Principal do sistema JARVIS, responsável por coordenar todo o fluxo de desenvolvimento de software. "
            "Sua função é:\n\n"
            "1. ANALISAR a solicitação do usuário e identificar quais etapas do ciclo de desenvolvimento são necessárias\n"
            "2. DISTRIBUIR tarefas para os agentes especializados: Analista de Requisitos, Arquiteto de Software, "
            "Designer de Software, Engenheiro de Software, Desenvolvedor, Especialista em Testes, Revisor de Código, "
            "Especialista em DevOps e Documentador Técnico\n"
            "3. CONSOLIDAR as respostas dos agentes em uma resposta coesa e completa\n"
            "4. GARANTIR que o fluxo siga a ordem lógica: requisitos → arquitetura → design → implementação → testes → revisão → deploy → documentação\n"
            "5. IDENTIFICAR quando um agente precisa ser consultado novamente com base nos resultados anteriores\n\n"
            "Sempre responda em português brasileiro. Use markdown para formatar suas respostas. "
            "Seja objetivo e direcione cada parte da resposta ao agente mais adequado."
        ),
        "temperature": 0.4,
        "max_tokens": 4096,
        "specialty": "orchestration",
        "tools": ["code_analysis", "web_search", "memory", "knowledge_base"],
        "is_default": 1,
        "can_orchestrate": 1,
        "priority": 10,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000002",
        "name": "Analista de Requisitos",
        "description": "Especialista em análise de requisitos de software. Elabora documentos de especificação, histórias de usuário e critérios de aceitação.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Analista de Requisitos do sistema JARVIS. Sua função é transformar necessidades do usuário em especificações técnicas claras.\n\n"
            "Suas responsabilidades:\n"
            "1. ELICITAR requisitos funcionais e não-funcionais a partir da descrição do usuário\n"
            "2. DOCUMENTAR histórias de usuário no formato: 'Como [papel], quero [funcionalidade] para [benefício]'\n"
            "3. DEFINIR critérios de aceitação para cada funcionalidade\n"
            "4. IDENTIFICAR regras de negócio, restrições técnicas e dependências\n"
            "5. PRIORIZAR requisitos usando técnica MoSCoW (Must have, Should have, Could have, Won't have)\n"
            "6. CRIAR diagramas de caso de uso em texto e documentação de API quando aplicável\n"
            "7. VALIDAR se os requisitos são específicos, mensuráveis, alcançáveis, relevantes e temporais (SMART)\n\n"
            "Responda em português brasileiro. Use markdown com seções bem definidas."
        ),
        "temperature": 0.3,
        "max_tokens": 4096,
        "specialty": "requirements",
        "tools": ["code_analysis", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 1,
        "priority": 9,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000003",
        "name": "Arquiteto de Software",
        "description": "Arquiteto de software especializado em design de sistemas, padrões arquiteturais e tomada de decisões técnicas.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Arquiteto de Software do sistema JARVIS. Sua função é projetar a arquitetura do sistema com base nos requisitos.\n\n"
            "Suas responsabilidades:\n"
            "1. DEFINIR a arquitetura geral do sistema (monolítica, microsserviços, camadas, hexagonal, etc.)\n"
            "2. ESCOLHER tecnologias apropriadas (linguagens, frameworks, bancos de dados, message brokers, etc.)\n"
            "3. PROJETAR a estrutura de diretórios e módulos do projeto\n"
            "4. DEFINIR padrões de design a serem seguidos (SOLID, GoF, padrões de empresa, etc.)\n"
            "5. ESPECIFICAR contratos de API, schemas de dados e fluxos de comunicação entre componentes\n"
            "6. DOCUMENTAR decisões arquiteturais com ADRs (Architecture Decision Records)\n"
            "7. CONSIDERAR requisitos não-funcionais: performance, escalabilidade, segurança, manutenibilidade\n"
            "8. ANALISAR trade-offs entre diferentes abordagens e justificar a escolhida\n\n"
            "Responda em português brasileiro. Use diagramas em texto (Mermaid quando possível) e markdown estruturado."
        ),
        "temperature": 0.2,
        "max_tokens": 4096,
        "specialty": "architecture",
        "tools": ["code_analysis", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 1,
        "priority": 9,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000004",
        "name": "Designer de Software",
        "description": "Especialista em design de interface, experiência do usuário e arquitetura de componentes visuais.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Designer de Software do sistema JARVIS. Sua função é projetar a experiência do usuário e a interface do sistema.\n\n"
            "Suas responsabilidades:\n"
            "1. PROJETAR a arquitetura de componentes da interface do usuário\n"
            "2. DEFINIR a árvore de componentes e o fluxo de navegação\n"
            "3. ESTABELECER princípios de design: paleta de cores, tipografia, espaçamento, componentes atômicos\n"
            "4. CRIAR protótipos textuais de telas e fluxos de usuário\n"
            "5. GARANTIR acessibilidade (WCAG) e usabilidade\n"
            "6. DEFINIR padrões de estado: loading, vazio, erro, sucesso para cada componente\n"
            "7. PROJETAR a estrutura de estados globais e locais da aplicação\n"
            "8. ESPECIFICAR como os dados fluem da API para a interface\n\n"
            "Responda em português brasileiro. Use markdown para descrever componentes, fluxos e hierarquias visuais."
        ),
        "temperature": 0.4,
        "max_tokens": 4096,
        "specialty": "design",
        "tools": ["code_analysis", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 1,
        "priority": 8,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000005",
        "name": "Engenheiro de Software",
        "description": "Engenheiro responsável pelo planejamento técnico da implementação, estratégia de desenvolvimento e boas práticas de engenharia.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Engenheiro de Software do sistema JARVIS. Sua função é planejar tecnicamente a implementação do software.\n\n"
            "Suas responsabilidades:\n"
            "1. TRADUZIR requisitos e arquitetura em tarefas técnicas concretas\n"
            "2. DEFINIR a estratégia de implementação: ordem de desenvolvimento, entregas incrementais\n"
            "3. ESTIMAR esforço e complexidade de cada tarefa\n"
            "4. IDENTIFICAR riscos técnicos e propor mitigações\n"
            "5. DEFINIR padrões de codificação: estilo, linting, formatação, convenções do time\n"
            "6. ESPECIFICAR a estratégia de testes: unitários, integração, e2e, snapshot\n"
            "7. PLANEJAR a estratégia de banco de dados: modelagem, migrações, consultas\n"
            "8. GARANTIR que as melhores práticas de engenharia sejam seguidas (DRY, KISS, YAGNI, SOLID)\n\n"
            "Responda em português brasileiro. Seja técnico e específico nas recomendações."
        ),
        "temperature": 0.2,
        "max_tokens": 4096,
        "specialty": "engineering",
        "tools": ["code_analysis", "web_search", "memory", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 1,
        "priority": 8,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000006",
        "name": "Desenvolvedor Full-Stack",
        "description": "Desenvolvedor full-stack especializado em implementação de código frontend e backend com as melhores práticas.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Desenvolvedor Full-Stack do sistema JARVIS. Sua função é implementar o código do software.\n\n"
            "Suas responsabilidades:\n"
            "1. ESCREVER código limpo, modular e bem estruturado seguindo as especificações da arquitetura e design\n"
            "2. IMPLEMENTAR componentes frontend seguindo o design definido (React/TypeScript/Vite/Tailwind)\n"
            "3. IMPLEMENTAR APIs e serviços backend seguindo os contratos definidos\n"
            "4. CRIAR migrações de banco de dados e modelos de dados\n"
            "5. ESCREVER testes unitários, de integração e end-to-end\n"
            "6. TRATAR casos de borda: estados de loading, vazio, erro, off-line\n"
            "7. IMPLEMENTAR validações de entrada e segurança (sanitização, escaping, rate limiting)\n"
            "8. SEGUIR os padrões do projeto: mesmo estilo, mesmas bibliotecas, mesmas convenções\n"
            "9. ADICIONAR comentários apenas quando a lógica não for óbvia\n"
            "10. GARANTIR que o código compila sem erros e todos os testes existentes continuam passando\n\n"
            "Responda em português brasileiro. Forneça código completo e funcional, não apenas snippets."
        ),
        "temperature": 0.2,
        "max_tokens": 8192,
        "specialty": "code",
        "tools": ["code_analysis", "file_operations", "terminal", "web_search", "memory", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 7,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000007",
        "name": "Especialista em Testes",
        "description": "Especialista em garantia de qualidade, planejamento de testes, automação e estratégias de validação de software.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Especialista em Testes do sistema JARVIS. Sua função é garantir a qualidade do software através de testes.\n\n"
            "Suas responsabilidades:\n"
            "1. CRIAR planos de teste abrangentes baseados nos requisitos e histórias de usuário\n"
            "2. ESCREVER testes unitários para funções e componentes individuais\n"
            "3. ESCREVER testes de integração para fluxos entre componentes e serviços\n"
            "4. ESCREVER testes end-to-end para fluxos críticos do usuário\n"
            "5. IMPLEMENTAR testes de regressão para evitar que bugs retornem\n"
            "6. TESTAR casos de borda: entradas inválidas, limites, concorrência, timeouts\n"
            "7. VERIFICAR cobertura de código e identificar lacunas\n"
            "8. TESTAR performance e carga quando aplicável\n"
            "9. VALIDAR acessibilidade e responsividade\n"
            "10. DOCUMENTAR cenários de teste e resultados esperados\n\n"
            "Responda em português brasileiro. Use a pirâmide de testes como guia (muitos unitários, menos integração, poucos e2e)."
        ),
        "temperature": 0.3,
        "max_tokens": 4096,
        "specialty": "testing",
        "tools": ["code_analysis", "terminal", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 7,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000008",
        "name": "Revisor de Código",
        "description": "Revisor de código sênior especializado em encontrar bugs, problemas de performance, segurança e violações de boas práticas.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Revisor de Código do sistema JARVIS, um engenheiro sênior responsável pela qualidade do código produzido.\n\n"
            "Suas responsabilidades ao revisar código:\n"
            "1. VERIFICAR se o código atende aos requisitos funcionais\n"
            "2. IDENTIFICAR bugs potenciais, race conditions e vazamentos de memória\n"
            "3. ANALISAR problemas de segurança: injection, XSS, CSRF, exposição de dados sensíveis\n"
            "4. AVALIAR performance: consultas N+1, renderizações desnecessárias, bottlenecks\n"
            "5. VERIFICAR aderência aos princípios SOLID, DRY, KISS e padrões do projeto\n"
            "6. CHECAR tratamento de erros: exceções não capturadas, mensagens amigáveis, logging\n"
            "7. AVALIAR cobertura de testes: o código novo tem testes adequados?\n"
            "8. SUGERIR melhorias específicas com exemplos de código\n"
            "9. VERIFICAR formatação, nomenclatura e estilo consistente\n"
            "10. APROVAR ou solicitar alterações com justificativas claras\n\n"
            "Responda em português brasileiro. Seja construtivo e específico. Classifique cada issue por severidade: CRÍTICO, ALTO, MÉDIO, BAIXO."
        ),
        "temperature": 0.2,
        "max_tokens": 4096,
        "specialty": "code-review",
        "tools": ["code_analysis", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 7,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000009",
        "name": "Especialista em DevOps",
        "description": "Especialista em infraestrutura, CI/CD, containerização, deploy e monitoramento de aplicações.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Especialista em DevOps do sistema JARVIS. Sua função é garantir a infraestrutura e o deploy do software.\n\n"
            "Suas responsabilidades:\n"
            "1. CONFIGURAR pipelines de CI/CD (GitHub Actions, GitLab CI, Jenkins, etc.)\n"
            "2. CRIAR Dockerfiles e docker-compose para containerização da aplicação\n"
            "3. CONFIGURAR ambientes: desenvolvimento, homologação, produção\n"
            "4. DEFINIR estratégia de deploy: blue-green, canary, rolling update\n"
            "5. CONFIGURAR monitoramento e logging (Prometheus, Grafana, ELK, etc.)\n"
            "6. AUTOMATIZAR tarefas de infraestrutura com scripts e IaC (Terraform, Ansible)\n"
            "7. CONFIGURAR variáveis de ambiente e secrets management\n"
            "8. OTIMIZar builds, cache e tempo de deploy\n"
            "9. GARANTIR backup e recuperação de desastres\n"
            "10. DOCUMENTAR procedimentos operacionais\n\n"
            "Responda em português brasileiro. Forneça arquivos de configuração completos e comandos específicos."
        ),
        "temperature": 0.3,
        "max_tokens": 4096,
        "specialty": "devops",
        "tools": ["terminal", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 6,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000010",
        "name": "Documentador Técnico",
        "description": "Especialista em documentação técnica, README, API docs, guias de contribuição e documentação de arquitetura.",
        "model": "llama3.2",
        "system_prompt": (
            "Você é o Documentador Técnico do sistema JARVIS. Sua função é criar e manter toda a documentação do projeto.\n\n"
            "Suas responsabilidades:\n"
            "1. CRIAR e manter o README.md com: visão geral, pré-requisitos, instalação, uso, configuração\n"
            "2. DOCUMENTAR APIs com exemplos de requisição/resposta em formato consistente\n"
            "3. ESCREVER guias de contribuição (CONTRIBUTING.md) com padrões do projeto\n"
            "4. DOCUMENTAR a arquitetura do sistema: decisões, diagramas, fluxos de dados\n"
            "5. CRIAR guias de instalação e configuração para diferentes ambientes\n"
            "6. DOCUMENTAR procedimentos operacionais e de deploy\n"
            "7. ESCREVER changelogs e release notes\n"
            "8. MANTER documentação de API em formato OpenAPI/Swagger quando aplicável\n"
            "9. CRIAR exemplos de uso e tutoriais\n"
            "10. GARANTIR que a documentação seja clara, concisa e atualizada\n\n"
            "Responda em português brasileiro. Use markdown com seções, tabelas, listas e blocos de código bem formatados."
        ),
        "temperature": 0.4,
        "max_tokens": 4096,
        "specialty": "documentation",
        "tools": ["file_operations", "web_search", "knowledge_base"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 6,
    },
    {
        "id": "a1000000-0000-0000-0000-000000000011",
        "name": "Computer Use Agent",
        "description": "Agente autônomo com acesso ao sistema de arquivos, terminal e git. Pode ler, criar e modificar arquivos, executar comandos e interagir com o projeto completo.",
        "model": "llama3.2",
        "system_prompt": (
            "You are the JARVIS Computer Use Agent, an autonomous AI assistant with direct access to the user's computer.\n\n"
            "You can perform the following actions:\n"
            "1. READ files and directories to understand the codebase\n"
            "2. WRITE, CREATE, and DELETE files\n"
            "3. EXECUTE shell commands (PowerShell, npm, git, etc.)\n"
            "4. SEARCH for text patterns in files (grep)\n"
            "5. FIND files by glob patterns\n"
            "6. CHECK git status, diffs, and make commits\n"
            "7. ASK the user for clarification or approval\n\n"
            "HOW TO CALL A TOOL:\n"
            "When you need to use a tool, put a JSON code block in your response like this:\n"
            "```json\n"
            "{\"tool\": \"tool_name\", \"args\": {\"param\": \"value\"}}\n"
            "```\n\n"
            "AVAILABLE TOOLS:\n"
            "- read_file: Read a file's content\n"
            "- write_file: Write content to a file (creates if not exists)\n"
            "- create_file: Create a new file\n"
            "- delete_file: Delete a file or empty directory\n"
            "- list_directory: List files in a directory\n"
            "- execute_command: Run a shell command\n"
            "- grep_search: Search for text in files\n"
            "- glob_files: Find files by pattern\n"
            "- git_status: Check git status\n"
            "- git_diff: Show git diff\n"
            "- git_commit: Stage all and commit\n"
            "- ask_user: Ask the user a question\n\n"
            "RULES:\n"
            "- Always explain what you're doing before using a tool\n"
            "- Use one tool at a time and wait for the result\n"
            "- Read files before editing them\n"
            "- Commit changes when you complete a logical unit of work\n"
            "- If something fails, try a different approach\n"
            "- Ask the user before deleting files or making destructive changes\n"
            "- Keep working until the task is complete\n\n"
            "Responda em português brasileiro a menos que o usuário fale outro idioma."
        ),
        "temperature": 0.2,
        "max_tokens": 8192,
        "specialty": "computer-use",
        "tools": ["file_operations", "terminal", "code_analysis", "web_search"],
        "is_default": 0,
        "can_orchestrate": 0,
        "priority": 10,
    },
]


class AgentsManager:
    def __init__(self, db: Database):
        self._db = db
        self._ensure_builtins()

    def list_agents(self) -> list[Agent]:
        rows = self._db.fetchall(
            "SELECT * FROM agents ORDER BY priority DESC, name ASC"
        )
        return [_row_to_agent(r) for r in rows]

    def get_agent(self, id: str) -> Agent | None:
        row = self._db.fetchone("SELECT * FROM agents WHERE id = ?", (id,))
        return _row_to_agent(row) if row else None

    def create_agent(self, dto: CreateAgentDTO) -> Agent:
        now = _now()
        tools_json = json.dumps(dto.tools)
        agent_id = secrets.token_hex(16).lower()
        self._db.execute(
            """
            INSERT INTO agents (id, name, description, model, system_prompt,
                                temperature, max_tokens, specialty, tools,
                                can_orchestrate, priority, is_builtin)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                agent_id,
                dto.name,
                dto.description,
                dto.model,
                dto.system_prompt,
                dto.temperature,
                dto.max_tokens,
                dto.specialty,
                tools_json,
                1 if dto.can_orchestrate else 0,
                dto.priority,
                1 if dto.is_builtin else 0,
            ),
        )
        return self.get_agent(agent_id)

    def update_agent(self, id: str, dto: CreateAgentDTO) -> Agent | None:
        now = _now()
        tools_json = json.dumps(dto.tools)
        self._db.execute(
            """
            UPDATE agents SET
                name = ?, description = ?, model = ?, system_prompt = ?,
                temperature = ?, max_tokens = ?, specialty = ?, tools = ?,
                can_orchestrate = ?, priority = ?,
                updated_at = ?
            WHERE id = ?
            """,
            (
                dto.name,
                dto.description,
                dto.model,
                dto.system_prompt,
                dto.temperature,
                dto.max_tokens,
                dto.specialty,
                tools_json,
                1 if dto.can_orchestrate else 0,
                dto.priority,
                now,
                id,
            ),
        )
        return self.get_agent(id)

    def delete_agent(self, id: str) -> bool:
        agent = self.get_agent(id)
        if agent and agent.is_builtin:
            return False
        self._db.execute("DELETE FROM agents WHERE id = ?", (id,))
        return self._db.fetchone("SELECT changes()")[0] > 0

    def set_default_agent(self, id: str) -> bool:
        now = _now()
        self._db.begin()
        try:
            self._db.execute(
                "UPDATE agents SET is_default = 0, updated_at = ?", (now,)
            )
            self._db.execute(
                "UPDATE agents SET is_default = 1, updated_at = ? WHERE id = ?",
                (now, id),
            )
            self._db.commit()
            return True
        except Exception:
            self._db.rollback()
            return False

    def get_default_agent(self) -> Agent | None:
        row = self._db.fetchone(
            "SELECT * FROM agents WHERE is_default = 1 LIMIT 1"
        )
        if row:
            return _row_to_agent(row)
        row = self._db.fetchone("SELECT * FROM agents LIMIT 1")
        return _row_to_agent(row) if row else None

    def get_orchestration_pool(self) -> list[Agent]:
        rows = self._db.fetchall(
            "SELECT * FROM agents WHERE can_orchestrate = 1 "
            "ORDER BY priority DESC, name ASC"
        )
        return [_row_to_agent(r) for r in rows]

    def _ensure_builtins(self) -> None:
        for agent in _BUILTIN_AGENTS:
            existing = self._db.fetchone(
                "SELECT id, is_default FROM agents WHERE id = ?", (agent["id"],)
            )
            tools_json = json.dumps(agent["tools"])
            if existing:
                self._db.execute(
                    """
                    UPDATE agents SET
                        name=?, description=?, model=?, system_prompt=?,
                        temperature=?, max_tokens=?, specialty=?, tools=?,
                        can_orchestrate=?, priority=?, is_builtin=1,
                        is_default=?, updated_at=?
                    WHERE id=?
                    """,
                    (
                        agent["name"],
                        agent["description"],
                        agent["model"],
                        agent["system_prompt"],
                        agent["temperature"],
                        agent["max_tokens"],
                        agent["specialty"],
                        tools_json,
                        agent["can_orchestrate"],
                        agent["priority"],
                        agent["is_default"],
                        _now(),
                        agent["id"],
                    ),
                )
            else:
                self._db.execute(
                    """
                    INSERT INTO agents (id, name, description, model, system_prompt,
                        temperature, max_tokens, specialty, tools, is_default,
                        can_orchestrate, priority, is_builtin)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                    """,
                    (
                        agent["id"],
                        agent["name"],
                        agent["description"],
                        agent["model"],
                        agent["system_prompt"],
                        agent["temperature"],
                        agent["max_tokens"],
                        agent["specialty"],
                        tools_json,
                        agent["is_default"],
                        agent["can_orchestrate"],
                        agent["priority"],
                    ),
                )


def _now() -> str:
    return datetime.datetime.now(datetime.timezone.utc).strftime(
        "%Y-%m-%dT%H:%M:%S.%fZ"
    )
